File: nlm_floating_point/PBCH/ps_nr_pbch_bit_descrambling.m
function bitOut  = ps_nr_pbch_bit_descrambling(bitIn, cellId, numSsbBlocks)

validNumSSB = [4 8 64];
bitOut = [];

if ((cellId > 1007) || (cellId < 0))
    disp("Invalid CellId");
end

if (~any(validNumSSB == numSsbBlocks))
    disp("Invalid Number of SSB");

end

pbchPayloadInterleaverPattern = [16,23,18,17,8,30,10,6,24,7,0,5,3,2,1,4,9,11,12,13,14,15,19,20,21,22,25,26,27,28,29,31];
pbchPayloadSize = 32;

halfIdxBitPos =  pbchPayloadInterleaverPattern(1, 10 + 1);
sfnLsb2BitPos =  pbchPayloadInterleaverPattern(1, 8 + 1);
sfnLsb3BitPos =  pbchPayloadInterleaverPattern(1, 7 + 1);

if(numSsbBlocks == 64)
    M = pbchPayloadSize - 6;
else
    M = pbchPayloadSize - 3;
end

pnLength = 4 * M;
cInit    = cellId;
c = ps_nr_pseudo_random_generator(cInit, pnLength);
v = 2*bitIn(sfnLsb3BitPos + 1) + bitIn(sfnLsb2BitPos + 1);

j = 1;
for i = 1:1:pbchPayloadSize
    if((i == halfIdxBitPos + 1) || (i == sfnLsb2BitPos + 1) || (i == sfnLsb3BitPos + 1))
        bitOut(i, 1) = bitIn(i);
    else
        bitOut(i, 1) = mod((bitIn(i) + c(j + v*M)), 2);
        j = j + 1;
    end
end

end

File: nlm_controller/Demo/CORTEX_M4_NXP_LA9310_GCC/ps_nr_pbch_bit_descrambling.c

#include "FreeRTOS.h"
#include "task.h"
#include "debug_console.h"
#include "pbch.h"
#include "math.h"

void ps_nr_pbch_bit_descrambling(uint32_t* bitIn, uint8_t numSsbBlocks, uint8_t* pScrCw)
{
    uint8_t  pbchPayloadSize = 32;
    uint8_t  halfIdxBitPos  =  0;
    uint8_t  sfnLsb2BitPos  =  24;
    uint8_t  sfnLsb3BitPos  =  6;
    uint8_t  M, v;
    uint32_t temp, oneBitShift;
   
    if(numSsbBlocks == 64)
        M = pbchPayloadSize - 6;
    else
        M = pbchPayloadSize - 3;
        
    uint8_t i, j =0;
  
    v = (((*bitIn >> sfnLsb3BitPos) & 0x1) << 1) + ((*bitIn >> sfnLsb2BitPos) & 0x1);
    oneBitShift = *bitIn;
    *bitIn = 0;
    
    for(i = 0; i < pbchPayloadSize; i++)
    {
        if((i != halfIdxBitPos) && (i != sfnLsb2BitPos) && (i != sfnLsb3BitPos))
        {
            temp = ((oneBitShift & 0x1) + pScrCw[j+v*M]) & 0x1;
            temp = (temp << (i*temp));
            *bitIn = (*bitIn | temp);
            j = j+1;	
	    }
	    oneBitShift = oneBitShift >> 1;
    }
}